homeassistant/components/ozw/services.py

```diff
@@ -105,6 +105,7 @@ def async_set_config_parameter(self, service):
             or value.type == ValueType.BYTE
             or value.type == ValueType.SHORT
         ):
+            selection = int(selection)  # sanitize
```

Comment:
The service schema validation should already have converted the `selection` to an integer. I don't understand why this is needed.

Reply:
The schema validation allows strings:

https://github.com/home-assistant/core/blob/a514cd16e12cd79af5aa2854b42eafe780ed43d0/homeassistant/components/ozw/services.py#L56

We were already converting to an int at the payload, I just moved it up to the top.

Reply:
The problematic input was a template.

Reply:
If the rendered template can be coerced to an integer it should already have been done so by the service schema.

Reply:
The templates seem to be treated differently, at least that's what I am observing.

This works fine: `value : "2"`.
This gives the error: `value: "{{ 2 }}"`.

Reply:
Aside from the template issue, a user could make a mistake and input a string for an integer parameter. I would say the casting error is slightly more user-friendly than the comparison error.
```
>>> "a" < 2
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '<' not supported between instances of 'str' and 'int'
>>> int("a")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid literal for int() with base 10: 'a'
```
Still, catching the exception and returning a human-readable error message would be even better.

Reply:
Looks like the mentioned PR #40527 has already implemented this.

Reply:
OK, I had to look a bit more, and when using Dev Tools, the template code is not evaluated. I added a debug log and using `value: "{{ 2 }}"` I get:
```
[homeassistant.components.ozw.services] selection {{ 2 }} is python type <class 'str'>
```
Perhaps this is common knowledge. I've been using template data for this service call in automations without any issue, so the difference escaped me. I guess automations evaluate the template before it gets to the service call.  The original poster did not specify whether they were using an automation or the Dev Tools page, so I have to presume the latter now.

Can the service call support this by including `cv.template` as a validator?

Reply:
We should catch the error and raise an error with an informative error argument.

I'm still not sure why the service validation doesn't work as expected. In my simple interpreter tests the rendered template is validated correctly.

```
Python 3.7.9 (default, Aug 18 2020, 06:22:45) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.18.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import voluptuous as vol

In [2]: schema = vol.Schema({"value": vol.Any(vol.Coerce(int), str)})

In [3]: import jinja2

In [4]: template = jinja2.Template("{{ 2 }}")

In [5]: rendered = template.render()

In [6]: type(schema({"value": rendered})["value"])
Out[6]: int
```

Reply:
Your example is rendering the template prior to passing it to the schema validator (as does automations). Apparently Dev Tools -> Services does not render the template and just passes the string as-is. I couldn't find any documentation about it, but apparently it's [always](https://community.home-assistant.io/t/help-with-automation-using-templated-json/136628/12) been that way?

Based on that behavior, at best this or the other PR can only provide better error messages for OP's problem.